### PR TITLE
journal: adapt grpc journal into new abstraction 

### PR DESCRIPTION
**Pull request by tisonkun**
This is a follow up to #153 .

### CODE REVIEW

**Comment by tisonkun** [comment]
@huachaohuang gRPC journal fixed.

File: src/journal/src/grpc/journal.rs [hunk 1]
```diff
@@ -12,38 +12,34 @@
...
-
 use super::{
     client::Client,
-    error::Result,
     proto::{CreateStreamRequest, DeleteStreamRequest},
-    stream::RemoteStream,
+    stream::Stream,
 };
-use crate::{async_trait, Journal, Timestamp};
+use crate::{async_trait, error::Result};
```
Comment (huachaohuang) [inline src/journal/src/grpc/journal.rs:20]: I am OK with `crate::Result` now, so you may consider it :)

File: src/journal/src/grpc/journal.proto [hunk 2]
```diff
@@ -65,6 +65,10 @@ message ReadEventRequest {
 }
 
 message ReadEventResponse {
```
Comment (huachaohuang) [inline src/journal/src/grpc/journal.proto:67]: s/ReadEvent/ReadEvents, let's keep it consistent with the trait.

File: src/journal/src/grpc/proto.rs [hunk 3]
```diff
@@ -14,13 +14,12 @@
 
 tonic::include_proto!("engula.journal.grpc.v1");
```
Comment (huachaohuang) [inline src/journal/src/grpc/proto.rs:15]:
```suggestion
tonic::include_proto!("engula.journal.v1");
```

I think "engula.journal.v1" is sufficient, the "grpc" is redundant.

File: src/journal/src/grpc/server.rs [hunk 4]
```diff
@@ -12,61 +12,39 @@
...
-impl<S, J, T> Server<S, J, T>
+impl<J> Server<J>
 where
-    T: Timestamp + 'static,
-    S: Stream<Timestamp = T> + Send + Sync + 'static,
-    S::Error: Send + Sync + 'static,
-    S::EventStream: Send + Sync + 'static,
-    J: Journal<S> + Send + Sync + 'static,
-    Status: From<S::Error>,
+    J: Journal + Send + Sync + 'static,
```
Comment (huachaohuang) [inline src/journal/src/grpc/server.rs:30]: You can simply `impl<J: Journal> Server<J>`, `Journal` already implies `Send + Sync + 'static` now.

File: src/journal/src/grpc/server.rs [hunk 5]
```diff
@@ -12,61 +12,39 @@
...
-impl<S, J, T> journal_server::Journal for Server<S, J, T>
+impl<J> journal_server::Journal for Server<J>
 where
-    T: Timestamp + 'static,
-    S: Stream<Timestamp = T> + Send + Sync + 'static,
-    S::Error: Send + Sync + 'static,
-    S::EventStream: Send + Sync + 'static,
-    J: Journal<S> + Send + Sync + 'static,
-    Status: From<S::Error>,
+    J: Journal + Send + Sync + 'static,
```
Comment (huachaohuang) [inline src/journal/src/grpc/server.rs:44]: ditto

File: src/journal/src/grpc/mod.rs [hunk 6]
```diff
@@ -13,35 +13,28 @@
...
+    use super::Server;
     use crate::*;
 
     #[tokio::test(flavor = "multi_thread")]
     async fn test() -> std::result::Result<(), Box<dyn std::error::Error>> {
         let listener = TcpListener::bind("127.0.0.1:0").await?;
         let local_addr = listener.local_addr()?;
         tokio::task::spawn(async move {
-            let j: mem::MemJournal<u64> = mem::MemJournal::default();
+            let j: mem::Journal = mem::Journal::default();
```
Comment (huachaohuang) [inline src/journal/src/grpc/mod.rs:37]:
```suggestion
            let j = mem::Journal::default();
```

File: src/journal/src/grpc/proto.rs [hunk 7]
```diff
@@ -14,13 +14,12 @@
...
 
-use super::error::Error;
-use crate::Timestamp;
+use crate::{error::Error, Timestamp};
 
-pub fn serialize_ts<T: Timestamp>(ts: &T) -> Result<Vec<u8>, Error> {
+pub fn serialize_ts(ts: &Timestamp) -> Result<Vec<u8>, Error> {
```
Comment (huachaohuang) [inline src/journal/src/grpc/proto.rs:19]: Serializing a timestamp to json really bugs me and also makes the related code complicated. Let me think about it.

Comment (huachaohuang) [reply src/journal/src/grpc/proto.rs:19]: Maybe we consider it later. We can keep it this way for now to avoid mixing things in the same PR.

Comment (tisonkun) [reply src/journal/src/grpc/proto.rs:19]: Yep. Here is an abstraction saying we should serialize the timestamp. How to do it effectively can be considered. I'm unfamiliar with serde so it takes some time to get used to it and find an optimization.

**Comment by huachaohuang** [comment]
BTW, you can add the grpc implementation [here](https://github.com/engula/engula/blob/main/src/journal/src/lib.rs#L27).

File: src/journal/src/error.rs [hunk 8]
```diff
@@ -23,6 +23,26 @@ pub enum Error {
     AlreadyExists(String),
     #[error("{0}")]
     InvalidArgument(String),
+    #[error(transparent)]
+    GrpcStatus(#[from] tonic::Status),
```
[the remark this answers is not in the document]
Comment (tisonkun) [reply src/journal/src/error.rs:27]: Yes. I'm thinking of it. But what general errors should we name it? A fallback is `Internal`/`Unknown`, but I don't get a good idea.

[the remark this answers is not in the document]
Comment (huachaohuang) [reply src/journal/src/error.rs:27]: I think we can simply convert everything we haven't handled explicitly to `Unknown` for now.

[the remark this answers is not in the document]
Comment (tisonkun) [reply src/journal/src/error.rs:27]: Updated.

File: src/journal/src/grpc/journal.rs (hunk 1 above)
Comment (tisonkun) [reply src/journal/src/grpc/journal.rs:20]: Updated.

File: src/journal/src/grpc/journal.proto (hunk 2 above)
Comment (tisonkun) [reply src/journal/src/grpc/journal.proto:67]: Updated.

File: src/journal/src/grpc/proto.rs (hunk 3 above)
Comment (tisonkun) [reply src/journal/src/grpc/proto.rs:15]: OK.

**Comment by tisonkun** [comment]
@huachaohuang comments addressed.

**Review by huachaohuang — COMMENTED** [review]
Thanks, it looks much better now. One more thing, can you add a simple description like [this](https://github.com/engula/engula/blob/main/src/journal/src/lib.rs#L26) and [this](https://github.com/engula/engula/blob/main/src/journal/src/mem/mod.rs#L15)?

**Comment by tisonkun** [comment]
@huachaohuang updated.

**Review by huachaohuang — APPROVED** [review]

**Pull request merged**